### ISSUE TITLE
forward gives domain error when back is called on Float 16.

### ISSUE THREAD

**Issue by thebhatman** [FluxML/Zygote.jl#228]
This simple code breaks as it gives domain error when back is called on Float16.
```
using Zygote

using Flux

y = rand(Float16, 2)

y1 = rand(Float16, 2)

fwd, back = Zygote.forward(Flux.mse, y1, y)

back(one(Float16))
```
This is the error message :  
```
ERROR: DomainError with -0.23730469:
log will only return a complex result if called with a complex argument. Try log(Complex(x)).
```
But the same piece of code works when I use Float32 or Float64 or Int64. The problem is only with Float16.

**Comment by MikeInnes** [comment]
Can you try to narrow this down a bit? At a guess, it's probably due to the `^` function.

**Comment by thebhatman** [comment]
I just looked up the definitions of the log function in math.jl. I think the error is arising because log is only specifically defined for Float64 and Float32. The reason is that the StackTrace of the the error I get when calling back on Float16 is this : 
```
Stacktrace:
 [1] throw_complex_domainerror(::Symbol, ::Float32) at .\math.jl:31
 [2] log(::Float32) at .\special\log.jl:321
 [3] log at .\math.jl:1018 [inlined]
 [4] (::getfield(Zygote, Symbol("##584#585")){Float16,Int64})(::Float16) at C:\Users\manju\.julia\packages\Zygote\VeaFW\src\lib\number.jl:27
 [5] #1580#back at C:\Users\manju\.julia\packages\Zygote\VeaFW\src\lib\grad.jl:46 [inlined]
```

**Comment by thebhatman** [comment]
Yeah I observed that the error is mainly when `^` has negative input.  The same error can be reproduced in this manner : 
```
using Zygote

f(x) = x .^ 2

fwd, back = Zygote.forward(f, Float16(-2))

back(one(Float16))
```
Even this throws the domain error when called on Float16, but works perfectly for Float32 and Float64.

**Comment by MikeInnes** [comment]
Our rule for `^` comes from diffrules here: https://github.com/JuliaDiff/DiffRules.jl/blob/a9b00c81ac0d103888b124b2b9f58a7c815cebf9/src/rules.jl#L87

Would be great if you can show the Float16/Float32 difference outside of Zygote.

**Comment by thebhatman** [comment]
Yes, I found the cause of the error. It is in the Julia Base, specifically in this line https://github.com/JuliaLang/julia/blob/f6049d6d25d720bc90d82aadae6be98bc948ef24/base/intfuncs.jl#L235, Float 16 is not included as a type in this Union. So power is taking a different definition, which is causing it to use the second rule in Diffrule (i.e it is trying to calculate derivative wrt to exp rather than base) in the line you mentioned above.

**Comment by thebhatman** [comment]
If we add Float16 in the Union for HWReal, it solves the issue. This issue is not with Zygote.

**Comment by MikeInnes** [comment]
Where is HWReal used, that causes this difference? Even if it's not a Zygote issue we'll most likely have to find a workaround for this; probably easiest just to add some indirection / a dispatch one whatever is actually throwing the error.

**Comment by thebhatman** [comment]
I mean to say, that `^` is taking a different definition for Float32 and for Float16 because Float16 doesn't belong to HWReal. This different definition is causing it to use the diffrule ` ($x^$y) * log($x)` instead of the correct rule `$y * ($x^($y - 1))`. The difference in definitions is here https://github.com/JuliaLang/julia/blob/f6049d6d25d720bc90d82aadae6be98bc948ef24/base/intfuncs.jl#L244

**Comment by MikeInnes** [comment]
I see. Let's hack it for now then; just add the definition for `literal_pow` with Float16.

**Comment by thebhatman** [comment]
In Zygote? Should I add it Here : https://github.com/FluxML/Zygote.jl/blob/43c215c84c2e483708684f8631516570e277df50/src/lib/number.jl#L39 ?

**Comment by MikeInnes** [comment]
Sure, that seems fine.

**Comment by CarloLucibello** [comment]
The Float16 `DomainError` is gone; the MSE pullback returns valid Float16 gradients. Closing as fixed.

_(Triaged against Zygote v0.7.10 on Julia 1.12.6.)_
